### DISCUSSION TITLE
Pixelated Cooja UI -- Contiki 4.9 container on Ubuntu 22.04 on VMware Fusion 12 on macOS

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2646] [Q&A] [answered]
Configuration: Latest Contiki (Contiki-NG-288b134d6-dirty), in a Contiki container, running on Ubuntu 22.04, on VMware Fusion, on macOS Montery (12.6.7).

Cooja GUI is pixelated in this config. The buttons are not visible, compilation results are difficult to see, and so on.

More info: I have another Fusion VM on this computer where Cooja GUI is fine. However, it runs Contiki 4.7 on Ubuntu 20.04.

<img width="699" alt="Screen Shot 2023-08-30 at 15 42 37" src="https://github.com/contiki-ng/contiki-ng/assets/2912950/f4f46414-916d-4120-8c81-b5277252ac0b">

**Comment by pjonsson** [comment, marked as answer]
It doesn't happen on native macOS, so I'm guessing it's something with the graphics drivers of VMWare Fusion and miscalculations of what needs to be redrawn. It's not virtualization in itself, I use Virtualbox on Windows, and don't have the same problems.

You can try https://github.com/contiki-ng/cooja/pull/382 that uses a different look and feel. The PR also adds a command line switch so you can easily try a few other looks and feels and see if the bug happens for them too. (I don't appreciate the older looks from an aesthetic perspective, but a broken UI is far worse than a UI that looks dated.)

Edit: After writing the response, I realized that a change from 4.8 to 4.9 is that Cooja draws everything from the AWT thread, so graphics-related things should be strictly better.

**Reply by rahav** [reply]
1. Is there a container that runs on macOS?
2. I fetched the branch, but I don't see the parameter to use in the `--help` or what are the options

**Reply by pjonsson** [reply]
The docker image is x86-64 if that is what you're asking. I don't use a Mac, and I don't have a license for Docker desktop, but my impression was that M1/M2 were quite good at running Intel code through some built-in translation software.

You might be able to remove some Intel-packages from `tools/docker/Dockerfile` and rebuild the image locally for your architecture, `docker build tools/docker -t contiker/contiki-ng --no-cache --pull` is how the CI builds the image. If you succeed in building, I don't know what will happen with Linux/aarch64 though, Cooja isn't tested with that so it might not work at all.

You need `--args=--help` if you're using the cooja script in the container, otherwise you get Gradle help. The option name is `--look-and-feel`.

**Reply by rahav** [reply]
My preference is Ubuntu. In that container, I still don't see the `--look-and-feel` option:

````
user@4eff0244ee15:~/contiki-ng/examples/hello-world$ git branch
  develop
* use-flatlaf
user@4eff0244ee15:~/contiki-ng/examples/hello-world$ cooja --args=--version

> Task :run
WARNING: Using incubator modules: jdk.incubator.foreign
Cooja 4.8, Contiki-NG build interface version 2022071901
JVM: 17.0.7 (GraalVM Community OpenJDK 64-Bit Server VM 17.0.7+7-jvmci-23.0-b12)
OS: Linux 6.2.0-31-generic amd64

BUILD SUCCESSFUL in 2s
3 actionable tasks: 1 executed, 2 up-to-date
````

but
````
user@4eff0244ee15:~/contiki-ng/examples/hello-world$ cooja --args=--help

> Task :run
WARNING: Using incubator modules: jdk.incubator.foreign
Usage: <main class> [--[no-]gui] [--[no-]log-color] [--logdir=DIR]
                    [--contiki=DIR] [--cooja=DIR] [--javac=FILE]
                    [--config=FILE] [--random-seed=SEED] [--autostart]
                    [--update-simulation] [--print-contiki-ng-build-version]
                    [--print-simulation-config-version] [--platform=ARCH]
                    [--version] [--help] [FILE...]
      [FILE...]             one or more simulation files
      --[no-]gui            use graphical mode
      --[no-]log-color      use color in console log
      --logdir=DIR          the log directory use
      --contiki=DIR         the Contiki-NG directory
      --cooja=DIR           the Cooja directory
      --javac=FILE          the javac binary
      --config=FILE         the filename for external user config
      --random-seed=SEED    the random seed
      --autostart           automatically start simulations
      --update-simulation   write an updated simulation file (.csc) and exit
      --print-contiki-ng-build-version
                            print Contiki-NG build version
      --print-simulation-config-version
                            print simulation config version
      --platform=ARCH       MSPSim platform
      --version             print version information and exit
      --help                display a help message

BUILD SUCCESSFUL in 2s
3 actionable tasks: 1 executed, 2 up-to-date
````

**Reply by rahav** [reply]
Just wanted to let you know that this is not urgent in any way.

**Reply by pjonsson** [reply]
The local rebuild of the Docker image would just get an ARM version of Ubuntu (which I assume exists). I don't know if Cooja would work with that though, we have only tested on Linux/x86-64 and macOS (both Intel and ARM).

There's something with the checkout of the PR, first parameter from the top for me is listed.
```
$ cooja --args=--help
...
> Task :run
Usage: <main class> [--look-and-feel=LookAndFeel] [--[no-]gui] [--[no-]
                    log-color] [--logdir=DIR] [--contiki=DIR] [--cooja=DIR]
                    [--javac=FILE] [--config=FILE] [--random-seed=SEED]
                    [--autostart] [--update-simulation]
                    [--print-contiki-ng-build-version]
                    [--print-simulation-config-version] [--platform=ARCH]
                    [--version] [--help] [FILE...]
      [FILE...]             one or more simulation files
      --look-and-feel=LookAndFeel
                            one of: CrossPlatform, FlatLaf, Nimbus, System
      --[no-]gui            use graphical mode
      --[no-]log-color      use color in console log
      --logdir=DIR          the log directory use
```

I just pushed an update so the PR is rebased on latest master, and uses the latest version of FlatLaf. I usually don't keep the branches i check out from others, so I use the third answer with FETCH_HEAD when I want to try a PR without having to clean up after it:

https://stackoverflow.com/questions/27567846/how-can-i-check-out-a-github-pull-request-with-git

**Reply by pjonsson** [reply]
Having slept on things, I realized it could also be something with how GraalVM chooses to redraw things. It might be worth trying with a different JDK, and see if the problem persists. The image just unpacks GraalVM to /usr/local so GraalVM can't be uninstalled easily, but if you remove the java and javac files that might be enough for testing purposes. You can then install the openjdk-17-jdk package with apt.

If you try the master branch of Cooja, that uses Java 20 (until Java 21 is released). Gradle will download some JDK and I'm not sure which, but even if it is GraalVM, it's at least a different version than the one you are currently using.

**Reply by rahav** [reply]
Interesting. I fetched the new push from https://github.com/pjonsson/cooja.git (I assumed it is the same repo as https://github.com/contiki-ng/cooja/pull/382). Is that where you pushed it? Because I am not getting the new look-and-feel option.

I will try your suggestions with java as well.

**Reply by pjonsson** [reply]
Yes, it's the same branch, and the option is there, otherwise it wouldn't be visible in the "changed files" tab in the PR. Make sure that the branch you fetched is what you have checked out, and make sure that everything is rebuilt (the cooja script uses Gradle, so that shouldn't be the issue).
